actors/builtin/miner/miner_state.go

```diff
@@ -957,6 +991,140 @@ func (st *State) checkPrecommitExpiry(store adt.Store, sectors abi.BitField) (de
+
+		powerDelta, undeclaredFaultyPower, err = deadline.ProcessDeadlineEnd(store, quant, faultExpiration)
+		if err != nil {
+			return penalty, pledgeDelta, powerDelta, xerrors.Errorf("failed to process end of deadline %d: %w", dlInfo.Index, err)
+		}
+		totalFaultyPower = deadline.FaultyPower
+	}
+	{
+		// Expire sectors that are due, either for on-time expiration or "early" faulty-for-too-long.
+		expired, err := deadline.PopExpiredSectors(store, dlInfo.Last(), quant)
+		if err != nil {
+			return penalty, pledgeDelta, powerDelta, xerrors.Errorf("failed to load expired sectors: %w", err)
+		}
+
+		// Release pledge requirements for the sectors expiring on-time.
+		// Pledge for the sectors expiring early is retained to support the termination fee that will be assessed
+		// when the early termination is processed.
+		pledgeDelta = big.Sub(pledgeDelta, expired.OnTimePledge)
+		st.AddInitialPledgeRequirement(expired.OnTimePledge.Neg())
+
+		// Record reduction in power of the amount of expiring active power.
+		// Faulty power has already been lost, so the amount expiring can be excluded from the delta.
+		powerDelta = powerDelta.Sub(expired.ActivePower)
+
+		// Record deadlines with early terminations. While this
+		// bitfield is non-empty, the miner is locked until they
+		// pay the fee.
+		noEarlyTerminations, err := expired.EarlySectors.IsEmpty()
+		if err != nil {
+			return penalty, pledgeDelta, powerDelta, xerrors.Errorf("failed to count early terminations: %w", err)
+		}
+		if !noEarlyTerminations {
+			st.EarlyTerminations.Set(dlInfo.Index)
+		}
+
+		// The termination fee is paid later, in early-termination queue processing.
+		// We could charge at least the undeclared fault fee here, which is a lower bound on the penalty.
+		// https://github.com/filecoin-project/specs-actors/issues/674
+
+		// The deals are not terminated yet, that is left for processing of the early termination queue.
+	}
+
+	// Save new deadline state.
+	err = deadlines.UpdateDeadline(store, dlInfo.Index, deadline)
+	if err != nil {
+		return penalty, pledgeDelta, powerDelta, xerrors.Errorf("failed to update deadline %d: %w", dlInfo.Index, err)
+	}
+
+	err = st.SaveDeadlines(store, deadlines)
+	if err != nil {
+		return penalty, pledgeDelta, powerDelta, xerrors.Errorf("failed to save deadlines: %w", err)
+	}
+
+	// Compute penalties all together.
+	// Be very careful when changing these as any changes can affect
+	// rounding.
+	undeclaredPenalty := PledgePenaltyForUndeclaredFault(
+		rewardEstimate,
+		networkQAPowerEstimate,
+		undeclaredFaultyPower.QA,
+	)
+	ongoingPenalty := PledgePenaltyForDeclaredFault(
+		rewardEstimate,
+		networkQAPowerEstimate,
+		totalFaultyPower.QA,
+	)
+	// Compute the overlap so we can avoid double-charging undeclared faulty
+	// power as declared faulty power.
+	overlapPenalty := PledgePenaltyForDeclaredFault(
+		rewardEstimate,
+		networkQAPowerEstimate,
+		undeclaredFaultyPower.QA,
+	)
```

Comment:
I'd prefer to have compute `undeclaredFaultyPower` and `declaredFaultyPower` (`totalFaultyPower-undeclaredFaultyPower`) up-front, then compute two penalties. However, that changes rounding. Was there a reason for doing this one way versus the other?

Reply:
No, computing power first is preferable.